### ISSUE TITLE
MediEvil Resurrection #Nintendoswitch

### ISSUE THREAD

**Issue by mikifantastik98** [hrydgard/ppsspp#12706]
the camera keeps spinning around and won't stop...
...this bug makes the game unplayable.

*Nintendo Switch /  standalone version
the bug is on every Nintendo Switch version so far

**Comment by spriteice** [comment]
+1. Would definitely love to play this game on my Switch.

**Comment by unknownbrackets** [comment]
Does this happen on mobile devices too (they also use the same jit)?  Does it happen on PC?  I think people have not reported this on other platforms, but just want to confirm.

-[Unknown]

**Comment by spriteice** [comment]
As far as I know, it is only on the switch version.

**Comment by hrydgard** [comment]
Could still be some kind of issue translating the analog controller values? maybe some kind of wraparound (although sounds odd...)

**Comment by m4xw** [comment]
Also fwiw, this is the only game this happens with to my knowledge.

**Comment by spriteice** [comment]
I don’t think it seems to be a control issue, playing with the analogue deadzones did not seem to help. It seems that the green targeting thing is following something that is not visible on screen

**Comment by Panderner** [comment]
I've played this game on Android no analog drifting on touch screen and controllers

**Comment by m4xw** [comment]
@Panderner armv7 or aarch64?

**Comment by Panderner** [comment]
> @Panderner armv7 or aarch64?

aarch64

**Comment by m4xw** [comment]
Camera spin is supposedly fixed https://streamable.com/4ntguf
Another tester said tho the camera is "too high" https://www.youtube.com/watch?v=J2pA9fgRafg
def some change on last rebase (dunno how its supposed to be tho)

**Comment by hrydgard** [comment]
Extremely weird stuff. Changing the title of the issue.

**Comment by ilikemyrealname** [comment]
I'm having the camera spin issue, too. I tried both the regular and the "+" downloads https://www.ppsspp.org

**Comment by unknownbrackets** [comment]
Hm, that download says v1.9.x which is still a fairly old version...

-[Unknown]

**Comment by m4xw** [comment]
According to a tester, this issue still persists in my WIP rebase on master from 2 days ago

**Comment by alfiones** [comment]
For me...camera spin issue too with Medievil Resurrection EUR and Switch PPSSPP v.1.9.3-478
Is there another new version for Switch?

**Comment by unknownbrackets** [comment]
Does its spin depend on where exactly the analog stick is positioned?

-[Unknown]

**Comment by alfiones** [comment]
> Does its spin depend on where exactly the analog stick is positioned?
> 
> -[Unknown]

Exactly, then the character begins to turn on itself

**Comment by unknownbrackets** [comment]
This might just be a deadzone issue.  @m4xw we discussed this, but for more detail:

Currently, the deadzone is applied at the [platform level](https://github.com/hrydgard/ppsspp/blob/63a94ceebb5f82c2aa5015c6bd514df191755ab5/Windows/XinputDevice.cpp#L172).  Values aren't sent at all if they're [unchanged within the deadzone](https://github.com/hrydgard/ppsspp/blob/63a94ceebb5f82c2aa5015c6bd514df191755ab5/Windows/XinputDevice.cpp#L285).  In part, this is because some platforms may [already apply some deadzone logic automatically](https://github.com/hrydgard/ppsspp/blob/63a94ceebb5f82c2aa5015c6bd514df191755ab5/Windows/DinputDevice.cpp#L185).

The PSP internal APIs games call to get the controller analog data do **not** apply any deadzone logic.  Instead, they simply [return an X and Y both in the range 0 - 255](https://github.com/hrydgard/ppsspp/blob/master/Core/HLE/sceCtrl.cpp#L193), and games are left to apply deadzone logic or not.

On the PSP I used for years, the analog nub is a bit loose, but it's more the "cap" that's loose.  This'll be hard to understand if you've never used one physically, but it's more of a coarse pad you can slide around a bit than actually an analog stick.  The pad on mine is just slightly loose, so you can nudge it slightly without the actual stick underneath moving.

This at times (rarely) caused my nub to reset back slightly off center, which in many games did nothing.  But in a few, it made me run left always or spun the camera.  Luckily, in my case I could just nudge it back to center, but it was always annoying when it happened.

It could still be that the hardware applies some deadzone but from tests I did years ago, it seemed like the values changed as soon as I nudged the stick even the tiniest bit.

Most likely the fix here is to apply a deadzone before calling `__CtrlSetAnalogX()`, etc.  In my experience, proper analog sticks don't reset to center as rigidly or consistently as those PDP analog nub pad things... even Switch drift problems aside.

-[Unknown]

**Comment by ilikemyrealname** [comment]
> This might just be a deadzone issue.

It even happens with all controllers disconnected and just using the on-screen controls.

**Comment by unknownbrackets** [comment]
But not on Android/PC with touch controls?

-[Unknown]

**Comment by Panderner** [comment]
> But not on Android/PC with touch controls?
> 
> -[Unknown]

Yes it does not occur for my android phone.

**Comment by unknownbrackets** [comment]
Well, probably not deadzone then - although might still be good to apply deadzone on Switch anyway.

I wonder if we're getting some event from rotation or something and sending it via NativeAxis, though.  Could be a bug with how the SDL joystick stuff is implemented on libnx, and us not checking for SDL_CONTROLLER_AXIS_INVALID or something.

-[Unknown]

**Comment by hrydgard** [comment]
definitely seems the most likely to be some weird input mapping issue than a CPU emulation issue, as the Switch has the same CPU architecture as devices where it works...

**Comment by Paolo4300** [comment]
> > Does its spin depend on where exactly the analog stick is positioned?
> > -[Unknown]
> 
> Exactly, then the character begins to turn on itself

It's 2024 and I am having the same problem with the camara on the switch, have you found a solution to this bug? @alfiones

**Comment by JoseAIG** [comment]
This issue is still present in the new version 1.17.1-1056. FW 18.1.0 AMS 1.7.1

**Comment by hrydgard** [comment]
@JoseAIG @Paolo4300 Can you try going into Settings / Controls / Calibrate Analog Stick, and set "Deadzone Radius" to something like 0.30 ? Also, check if the left little circle on that screen is behaving erratically.

**Comment by ilikemyrealname** [comment]
@hrydgard @antipirgos  Changing Deadline Radius to .30 and/or using EUR version of game had no positive effect on the camera spin. Issue persists.
